### PR TITLE
Feature/add milestone 2 and 3 friendship requirements

### PR DESCRIPTION
**Pull request by Bluette1**
##### [Feature/add milestone 2 and 3 friendship requirements](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/)
- Covers the following milestones
  - Social media website - Milestone: friendships v1
  - Social media website - Milestone: friendships v2
- Implement the required friendship features
  - Create the required models and associations
  - Add the necessary logic in the controllers
  - Modify the corresponding views 
- Add tests
- [Live demo](https://post-bk-app.herokuapp.com/)

### CODE REVIEW

File: app/views/users/show.html.erb [hunk 1]
```diff
@@ -5,5 +5,11 @@
...
     </ul>
-  </article>
-</section>
+    <% if @logged_in_user %>
+      <h3>Pending Friend Requests:</h3>
+      <ul class="posts">
+        <%= render 'pending_friend_requests' %>
+      </ul>
+    <% end %>
+    <%= render 'add_or_accept_or_reject_friend' %>
```
Comment (maelfosso) [inline app/views/users/show.html.erb]: Kindly remove logic from views. Please implemented them in views helpers.

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,

I hope you're doing well. Thank you for the job done so far.
However, there are some missing requirements

1. It's required, as you can read it [here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone2/postgre.png), to use **Postgresql** from the beginning of the project. Kindly set it up.

2. Please, remove logics from views. Implement them in views helpers

3. Please, the `FriendRequest` table is not needed. Only `Friendship` table with a **`status`** will be sufficient. Kindly fix that first. It's the more important.

4 Your webapp must have a button to invite user to friendship on users list page and on single user page, as indicated [here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png), however, it only appears in the User list page and misses in the Single user page. Kindly added it, please.

5. You did a really good job concerning tests. However, integration tests are missing. Kindly add them.
Access to page, no model creation, just navigate through each page, click on links, fill forms and submit them and check the results from the new page and your database.

Kindly make the suggested changes and submit a new code review request
Thanks

Mael FOSSO

File: app/models/user.rb [hunk 2]
```diff
@@ -9,4 +9,23 @@ class User < ApplicationRecord
   has_many :posts
   has_many :comments, dependent: :destroy
   has_many :likes, dependent: :destroy
+
+  has_many :friendships, foreign_key: :friend_id, class_name: 'Friendship', dependent: :destroy
+  has_many :friends, through: :friendships, source: :inverse_friend, dependent: :destroy
```
Comment (stratospherique) [inline app/models/user.rb]: #### Implementation :warning: 
* [ ] User model `has_many :friendships` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friend.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [ ] User model `has_many :inverted_friendships, class_name: "Friendship", foreign_key: "friend_id"` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friend.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great Progress @Bluette1 :clap: However, there is still some requirements that weren't implemented. Check :arrow_down: 

#### Behaviour (Requried) :warning: 
* [x] Posts timeline includes posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
- When sending a friend request, the posts of the other user appear in the timeline automatically even though he didn't accept the friend request yet.
#### Tests
* [x] Unit testing for some models is missing

### That's it :relieved: When finished, kindly submit a code re-review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

**Comment by Bluette1** [comment]
@maelfosso and @stratospherique, thank you very much for your review and feedback!

**Comment by Dandush03** [comment]
# Status: :rocket: Project Approved :rocket: 

### GOOD job applying the necesary changes. 
### Happy Coding :nerd_face: :computer:

**Comment by Bluette1** [comment]
@Dandush03, thank you very much for your review and feedback!

**Pull request merged**